### PR TITLE
Added autopickup rules based on material types.

### LINKED ISSUE #25086: Auto pickup based on materials
**Is your feature request related to a problem? Please describe.**  
Often when you are scavenging items in bulk it not because you want the item itself but rather want to disassemble it for crafting materials, automatically scavenging everything of a material for later disassembly and crafting seems like the kind of thing auto pickup would be great for but this is as far as i can figure out not currently possible.

**Describe the solution you'd like**  
The ability to ad m: rules to auto pickup for example m:kevlar m:nomex etc.

### PR DESCRIPTION
**Pull request by OzoneH3**
#### Summary
```SUMMARY: Features "Added autopickup rules based on material types."``` 

#### Purpose of change
Fixes #25086

### CODE REVIEW

**Comment by cake-pie** [comment]
Suggestion for additional granularity/control:
- `m:` corresponds to `item::made_of_any( material_ids )` behavior -- true if at least one item of `material_ids`  matches any material of the item
  - `m:steel` will match any item that has steel as one of their materials
  - `m:steel,iron,superalloy` equivalent to three separate rules `m:steel` `m:iron` `m:superalloy`
- `M:` corresponds to `item::only_made_of( material_ids )` behavior -- false if the item has at least one material not in the set or, no materials at all
  - `M:steel` will match any item that has steel as their only material
  - `M:steel,iron,superalloy` will match any item that has at least one of those three materials, and none outside the set.

**Comment by cake-pie** [comment]
`only_made_of` behavior is useful for collecting pure metal items for recycling:
- The metal compactor `t_recycler` requires pure metal inputs, i.e. it won't accept a copper pot with a wooden handle, so `M:copper` would be more useful than `m:copper` for that purpose.
- It also allows material substitution in some cases, e.g. it will accept a mixture of steel, hard steel and budget steel and compact them into steel, so `M:steel, budget steel, hard steel` is useful for that.

If you could implement this while you're at it I'd appreciate it, since that'd take an item off my to-do list. If not, no big deal -- this is a bit of a niche feature, and I can understand if it's out of scope of what you'd bargained for when you picked up the issue.

**Comment by OzoneH3** [comment]
Implemented it like that. There'll probably be quite a bit of confusion how it works i feel like -.-

File: src/auto_pickup.cpp [hunk 1]
```diff
@@ -515,49 +523,107 @@ bool auto_pickup::empty() const
...
+    char type = ' ';
+    std::vector<std::string> filter;
+    if( rule[1] == ':' ) {
+        type = rule[0];
+        filter = string_split( rule.substr( 2 ), ',' );
+    }
+
+    if( filter.empty() || materials.empty() ) {
+        return false;
+
```
Comment (cake-pie) [inline src/auto_pickup.cpp]: stray line

**Comment by cake-pie** [comment]
> probably be quite a bit of confusion how it works

I think the issue is that "made of any"/"only made of" are taken from function names, which are terse, and themselves had to be [heavily commented](https://github.com/CleverRaven/Cataclysm-DDA/blob/c55272496ef74777320a5deb030c95ab52052b86/src/item.h#L804-L815) to aid understanding.

More intuitive explanation could look something like:
```
Pickup based on item materials:
m:kevlar        matches items made of kevlar
M:copper        matches items made purely of copper
M:steel,iron    multiple materials allowed (OR search)
```

**Pull request merged**